src/gui/plugins/visualize_frustum/VisualizeFrustum.cc

```diff
@@ -0,0 +1,442 @@
+                                          !this->dataPtr->visualDirty)
+  {
+    this->dataPtr->frustumPose = worldPose(this->dataPtr->frustumEntity, _ecm);
+  }
+}
+
+//////////////////////////////////////////////////
+void VisualizeFrustum::OnTopic(const QString &_topicName)
+{
+  if (!this->dataPtr->topicName.empty() &&
+      !this->dataPtr->node.Unsubscribe(this->dataPtr->topicName))
+  {
+    gzerr << "Unable to unsubscribe from topic ["
+          << this->dataPtr->topicName <<"]" <<std::endl;
+  }
+
+  this->dataPtr->topicName = _topicName.toStdString();
+  std::lock_guard<std::mutex> lock(this->dataPtr->serviceMutex);
+
+  // Reset visualization
+  this->dataPtr->resetVisual = true;
+
+  // Create new subscription
+  if (!this->dataPtr->node.Subscribe(this->dataPtr->topicName,
+                            &VisualizeFrustum::OnScan, this))
+  {
+    gzerr << "Unable to subscribe to topic ["
+          << this->dataPtr->topicName << "]\n";
+    return;
+  }
+  this->dataPtr->visualDirty = false;
+}
+
+//////////////////////////////////////////////////
+void VisualizeFrustum::DisplayVisual(bool _value)
+{
+  std::lock_guard<std::mutex> lock(this->dataPtr->serviceMutex);
+  this->dataPtr->frustum->SetVisible(_value);
+  gzerr << "Frustum Visual Display " << ((_value) ? "ON." : "OFF.")
+        << std::endl;
+}
+
+/////////////////////////////////////////////////
+void VisualizeFrustum::OnRefresh()
+{
+  // Clear
+  this->dataPtr->topicList.clear();
+
+  // Get updated list
+  std::vector<std::string> allTopics;
+  this->dataPtr->node.TopicList(allTopics);
+  for (auto topic : allTopics)
+  {
+    std::vector<transport::MessagePublisher> publishers;
+    this->dataPtr->node.TopicInfo(topic, publishers);
+    for (auto pub : publishers)
+    {
+      if (pub.MsgTypeName() == "gz.msgs.LogicalCameraSensor")
+      {
+        this->dataPtr->topicList.push_back(QString::fromStdString(topic));
+        break;
+      }
+    }
+  }
+  if (this->dataPtr->topicList.size() > 0)
+  {
+    this->OnTopic(this->dataPtr->topicList.at(0));
+  }
+
+  this->TopicListChanged();
+}
+
+/////////////////////////////////////////////////
+QStringList VisualizeFrustum::TopicList() const
+{
+  return this->dataPtr->topicList;
+}
+
+/////////////////////////////////////////////////
+void VisualizeFrustum::SetTopicList(const QStringList &_topicList)
+{
+  this->dataPtr->topicList = _topicList;
+  this->TopicListChanged();
+}
+
+//////////////////////////////////////////////////
+void VisualizeFrustum::OnScan(const msgs::LogicalCameraSensor &_msg)
+{
+  std::lock_guard<std::mutex> lock(this->dataPtr->serviceMutex);
+  if (this->dataPtr->initialized)
+  {
+    this->dataPtr->msg = std::move(_msg);
+
+    this->dataPtr->frustum->SetNear(this->dataPtr->msg.near_clip());
+    this->dataPtr->frustum->SetFar(this->dataPtr->msg.far_clip());
+    this->dataPtr->frustum->SetFOV(this->dataPtr->msg.horizontal_fov());
```

Comment:
some changes to API names were made upstream:

```suggestion
    this->dataPtr->frustum->SetNearClipPlane(this->dataPtr->msg.near_clip());
    this->dataPtr->frustum->SetFarClipPlane(this->dataPtr->msg.far_clip());
    this->dataPtr->frustum->SetHFOV(this->dataPtr->msg.horizontal_fov());
```